### PR TITLE
[TySan] A Type Sanitizer (Runtime Library)

### PR DESCRIPTION
**Pull request by fhahn**
This patch introduces the runtime components of a type sanitizer: a sanitizer for type-based aliasing violations.

C/C++ have type-based aliasing rules, and LLVM's optimizer can exploit these given TBAA metadata added by Clang. Roughly, a pointer of given type cannot be used to access an object of a different type (with, of course, certain exceptions). Unfortunately, there's a lot of code in the wild that violates these rules (e.g. for type punning), and such code often must be built with -fno-strict-aliasing. Performance is often sacrificed as a result. Part of the problem is the difficulty of finding TBAA violations. Hopefully, this sanitizer will help.

For each TBAA type-access descriptor, encoded in LLVM's IR using metadata, the corresponding instrumentation pass generates descriptor tables. Thus, for each type (and access descriptor), we have a unique pointer representation. Excepting anonymous-namespace types, these tables are comdat, so the pointer values should be unique across the program. The descriptors refer to other descriptors to form a type aliasing tree (just like LLVM's TBAA metadata does). The instrumentation handles the "fast path" (where the types match exactly and no partial-overlaps are detected), and defers to the runtime to handle all of the more-complicated cases. The runtime, of course, is also responsible for reporting errors when those are detected.

The runtime uses essentially the same shadow memory region as tsan, and we use 8 bytes of shadow memory, the size of the pointer to the type descriptor, for every byte of accessed data in the program. The value 0 is used to represent an unknown type. The value -1 is used to represent an interior byte (a byte that is part of a type, but not the first byte). The instrumentation first checks for an exact match between the type of the current access and the type for that address recorded in the shadow memory. If it matches, it then checks the shadow for the remainder of the bytes in the type to make sure that they're all -1. If not, we call the runtime. If the exact match fails, we next check if the value is 0 (i.e. unknown). If it is, then we check the shadow for the remainder of the byes in the type (to make sure they're all 0). If they're not, we call the runtime. We then set the shadow for the access address and set the shadow for the remaining bytes in the type to -1 (i.e. marking them as interior bytes). If the type indicated by the shadow memory for the access address is neither an exact match nor 0, we call the runtime.

The instrumentation pass inserts calls to the memset intrinsic to set the memory updated by memset, memcpy, and memmove, as well as allocas/byval (and for lifetime.start/end) to reset the shadow memory to reflect that the type is now unknown. The runtime intercepts memset, memcpy, etc. to perform the same function for the library calls.

The runtime essentially repeats these checks, but uses the full TBAA algorithm, just as the compiler does, to determine when two types are permitted to alias. In a situation where access overlap has occurred and aliasing is not permitted, an error is generated.

Clang's TBAA representation currently has a problem representing unions, as demonstrated by the one XFAIL'd test. We'll update the TBAA representation to fix this, and at the same time, update the sanitizer.

As a note, this implementation does not use the compressed shadow-memory scheme discussed previously (http://lists.llvm.org/pipermail/llvm-dev/2017-April/111766.html). That scheme would not handle the struct-path (i.e. structure offset) information that our TBAA represents. I expect we'll want to further work on compressing the shadow-memory representation, but I think it makes sense to do that as follow-up work.

(This includes build fixes for Linux from Mingjie Xu)

Based on https://reviews.llvm.org/D32197.

Depends on #76260 (Clang support), #76259 (LLVM support)

### CODE REVIEW

File: compiler-rt/cmake/config-ix.cmake [hunk 1]
```diff
@@ -720,7 +726,7 @@ if(COMPILER_RT_SUPPORTED_ARCH)
 endif()
 message(STATUS "Compiler-RT supported architectures: ${COMPILER_RT_SUPPORTED_ARCH}")
 
-set(ALL_SANITIZERS asan;dfsan;msan;hwasan;tsan;safestack;cfi;scudo_standalone;ubsan_minimal;gwp_asan;asan_abi)
+set(ALL_SANITIZERS asan;dfsan;msan;hwasan;tsan;tysan,safestack;cfi;scudo_standalone;ubsan_minimal;gwp_asan;asan_abi)
```
Comment (davemgreen) [inline compiler-rt/cmake/config-ix.cmake:729]: ^ `tysan;`

Comment (fhahn) [reply compiler-rt/cmake/config-ix.cmake]: Thanks, updated!

**Comment by fhahn** [comment]
Added compiler-rt tests for various strict-aliasing violations from the bug tracker I found.

**Comment by tavianator** [comment]
I also needed 

```diff
diff --git a/compiler-rt/cmake/config-ix.cmake b/compiler-rt/cmake/config-ix.cmake
index ab13d8c03683..f480083231a1 100644
--- a/compiler-rt/cmake/config-ix.cmake
+++ b/compiler-rt/cmake/config-ix.cmake
@@ -677,6 +677,7 @@ else()
   filter_available_targets(PROFILE_SUPPORTED_ARCH ${ALL_PROFILE_SUPPORTED_ARCH})
   filter_available_targets(CTX_PROFILE_SUPPORTED_ARCH ${ALL_CTX_PROFILE_SUPPORTED_ARCH})
   filter_available_targets(TSAN_SUPPORTED_ARCH ${ALL_TSAN_SUPPORTED_ARCH})
+  filter_available_targets(TYSAN_SUPPORTED_ARCH ${ALL_TYSAN_SUPPORTED_ARCH})
   filter_available_targets(UBSAN_SUPPORTED_ARCH ${ALL_UBSAN_SUPPORTED_ARCH})
   filter_available_targets(SAFESTACK_SUPPORTED_ARCH
     ${ALL_SAFESTACK_SUPPORTED_ARCH})
```

to build the runtime on LInux.

**Comment by fhahn** [comment]
The latest version also includes a new test case `compiler-rt/test/tysan/constexpr-subobject.cpp` which had a false positive (IIUC the code should have no strict aliasing violation) + a fix

File: compiler-rt/lib/tysan/tysan.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  uptr Offset;
+};
+
+struct tysan_struct_type_descriptor {
+  uptr MemberCount;
+  struct {
+    struct tysan_type_descriptor *Type;
+    uptr Offset;
+  } Members[1]; // Tail allocated.
+  // char Name[]; // Tail allocated.
```
Comment (fmayer) [inline compiler-rt/lib/tysan/tysan.h:50]: commented out code left?

Comment (fhahn) [reply compiler-rt/lib/tysan/tysan.h:50]: Dropped, thanks!

**Comment by fhahn** [comment]
The Clang and LLVM parts are now ready to land, with just the compiler-rt bits pending approval.

There are a few test failures on Linux which don't fail on macOS, will look at them today probably

**Comment by fhahn** [comment]
Test failures on linux should be fixed now, they were due to debug paths using full paths on Linux and in one case a C++ struct type wasnt' de-mangled on Linux

**Comment by fhahn** [comment]
ping :)

File: compiler-rt/lib/tysan/tysan.cpp [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    internal_memmove(shadow_for(daddr), shadow_for(saddr), size * sizeof(uptr));
+}
+
+static const char *getDisplayName(const char *Name) {
+  if (Name[0] == '\0')
+    return "<anonymous type>";
+
+  // Clang generates tags for C++ types that demangle as typeinfo. Remove the
+  // prefix from the generated string.
+  const char TIPrefix[] = "typeinfo name for ";
```
Comment (fmayer) [inline compiler-rt/lib/tysan/tysan.cpp:46]: can't we make this a `const char*`?

File: compiler-rt/lib/tysan/tysan.cpp [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+  const char *DName = Symbolizer::GetOrInit()->Demangle(Name);
+  if (!internal_strncmp(DName, TIPrefix, sizeof(TIPrefix) - 1))
+    DName += sizeof(TIPrefix) - 1;
```
Comment (fmayer) [inline compiler-rt/lib/tysan/tysan.cpp:50]: strlen? the compiler should be able to optimize that.

File: compiler-rt/lib/tysan/lit.cfg [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+# Setup source root.
+config.test_source_root = os.path.dirname(__file__)
+
+# Setup default compiler flags used with -fsanitize=type option.
+clang_tysan_cflags = (["-fsanitize=type",
+                      "-mno-omit-leaf-frame-pointer",
+                      "-fno-omit-frame-pointer",
+                      "-fno-optimize-sibling-calls"] +
+                      [config.target_cflags] +
```
Comment (fmayer) [inline compiler-rt/lib/tysan/lit.cfg:16]: This looks wrong, `target_cflags` sounds like a list

File: compiler-rt/lib/tysan/tysan.cpp [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+    if (TD->Tag == TYSAN_STRUCT_TD) {
+      if (TD->Struct.MemberCount > 0)
+        TD = TD->Struct.Members[0].Type;
+      else
+        TD = nullptr;
+    } else if (TD->Tag == TYSAN_MEMBER_TD) {
+      TD = TD->Member.Access;
+    } else {
+      DCHECK(0);
```
Comment (fmayer) [inline compiler-rt/lib/tysan/tysan.cpp:94]: `DCHECK(false && "invalid enum value")`?

File: compiler-rt/lib/tysan/tysan.cpp (hunk 3 above)
Comment (fhahn) [reply compiler-rt/lib/tysan/tysan.cpp]: Done together with using `strlen` thanks

File: compiler-rt/lib/tysan/tysan.cpp (hunk 4 above)
Comment (fhahn) [reply compiler-rt/lib/tysan/tysan.cpp]: Updated, thanks

File: compiler-rt/lib/tysan/tysan.cpp (hunk 6 above)
Comment (fhahn) [reply compiler-rt/lib/tysan/tysan.cpp]: Done, thanks!

File: compiler-rt/lib/tysan/tysan.cpp [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+static void printTDName(tysan_type_descriptor *td) {
+  if (((sptr)td) <= 0) {
+    Printf("<unknown type>");
+    return;
+  }
+
+  switch (td->Tag) {
+  default:
+    DCHECK(0);
```
Comment (fmayer) [inline compiler-rt/lib/tysan/tysan.cpp:66]: as below

File: compiler-rt/test/tysan/lit.cfg.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+import os
+import platform
+import re
+
+import lit.formats
+
+# Get shlex.quote if available (added in 3.3), and fall back to pipes.quote if
+# it's not available.
+try:
+    import shlex
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:12]: `from shlex import quote as sh_quote`?

File: compiler-rt/test/tysan/lit.cfg.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+    sh_quote = shlex.quote
+except:
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:15]: except ImportError?

File: compiler-rt/test/tysan/lit.cfg.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    import pipes
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:16]: as above

File: compiler-rt/test/tysan/lit.cfg.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    sh_quote = pipes.quote
+
+
+def get_required_attr(config, attr_name):
+    attr_value = getattr(config, attr_name, None)
+    if attr_value == None:
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:23]: `is None`

File: compiler-rt/test/tysan/lit.cfg.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+def push_dynamic_library_lookup_path(config, new_path):
+    if platform.system() == "Windows":
+        dynamic_library_lookup_var = "PATH"
+    elif platform.system() == "Darwin":
+        dynamic_library_lookup_var = "DYLD_LIBRARY_PATH"
+    else:
+        dynamic_library_lookup_var = "LD_LIBRARY_PATH"
+
+    new_ld_library_path = os.path.pathsep.join(
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:40]: why not `os.path.join`

File: compiler-rt/test/tysan/lit.cfg.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        (new_path, config.environment.get(dynamic_library_lookup_var, ""))
+    )
+    config.environment[dynamic_library_lookup_var] = new_ld_library_path
+
+    if platform.system() == "FreeBSD":
+        dynamic_library_lookup_var = "LD_32_LIBRARY_PATH"
+        new_ld_32_library_path = os.path.pathsep.join(
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:47]: as above

File: compiler-rt/test/tysan/lit.cfg.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            (new_path, config.environment.get(dynamic_library_lookup_var, ""))
+        )
+        config.environment[dynamic_library_lookup_var] = new_ld_32_library_path
+
+    if platform.system() == "SunOS":
+        dynamic_library_lookup_var = "LD_LIBRARY_PATH_32"
+        new_ld_library_path_32 = os.path.pathsep.join(
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:54]: as above

File: compiler-rt/test/tysan/lit.cfg.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+            (new_path, config.environment.get(dynamic_library_lookup_var, ""))
+        )
+        config.environment[dynamic_library_lookup_var] = new_ld_library_path_32
+
+        dynamic_library_lookup_var = "LD_LIBRARY_PATH_64"
+        new_ld_library_path_64 = os.path.pathsep.join(
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:60]: as above

File: compiler-rt/test/tysan/lit.cfg.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+# Setup config name.
+config.name = "TypeSanitizer" + config.name_suffix
+
+# Platform-specific default TYSAN_OPTIONS for lit tests.
+default_tysan_opts = list(config.default_sanitizer_opts)
+
+# On Darwin, leak checking is not enabled by default. Enable on macOS
+# tests to prevent regressions
+if config.host_os == "Darwin" and config.apple_platform == "osx":
```
Comment (fmayer) [inline compiler-rt/test/tysan/lit.cfg.py:74]: this looks like a leftover? what does tysan have to do with leak checking?

File: compiler-rt/lib/tysan/tysan.cpp (hunk 7 above)
Comment (fhahn) [reply compiler-rt/lib/tysan/tysan.cpp]: Done thanks!

File: compiler-rt/test/tysan/lit.cfg.py (hunk 8 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py]: Removed the conditional import, as we currently require Python >= 3.8 https://llvm.org/docs/GettingStarted.html#software

File: compiler-rt/test/tysan/lit.cfg.py (hunk 9 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py]: Removed, thanks

File: compiler-rt/test/tysan/lit.cfg.py (hunk 10 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py]: Removed, thanks

File: compiler-rt/test/tysan/lit.cfg.py (hunk 11 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py]: Changed to check `if attr_value` and return value early.

File: compiler-rt/test/tysan/lit.cfg.py (hunk 12 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py:31]: This joins together paths for used in an environment variable, IIUC e.g. `os.path.join` will use `/` on macOS to join parts of a single path and `pathsep` will be `:` used to join together multiple multiple separate paths

File: compiler-rt/test/tysan/lit.cfg.py (hunk 13 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py:38]: ditto

File: compiler-rt/test/tysan/lit.cfg.py (hunk 15 above)
Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py:51]: ditto

**Review by fmayer — APPROVED** [review]

File: compiler-rt/test/tysan/lit.cfg.py (hunk 11 above)
Comment (fmayer) [reply compiler-rt/test/tysan/lit.cfg.py]: just confirming it is expected that empty string, or false, or something like that will now fail

Comment (fhahn) [reply compiler-rt/test/tysan/lit.cfg.py]: I've adjusted to return if not None, thanks

**Comment by kazutakahirata** [comment]
@fhahn I'm getting:

```
compiler-rt/lib/tysan/../sanitizer_common/sanitizer_platform_limits_posix.h:604:3: error: anonymous structs are a GNU extension [-Werror,-Wgnu-anonymous-struct]
  struct {
  ^
1 error generated.
```

while compiling `tyscan.cpp`.  Is there any way you could take a look?  Thanks!

**Comment by fhahn** [comment]
AFAICT this is happening at each place that includes the file, e.g. also if you build AddressSanitizer. Not sure what the best way forward here is as this is not related to TypeSanitizer, but seems to impact all sanitizers?

**Comment by kazutakahirata** [comment]
> AFAICT this is happening at each place that includes the file, e.g. also if you build AddressSanitizer. Not sure what the best way forward here is as this is not related to TypeSanitizer, but seems to impact all sanitizers?

I just posted https://github.com/llvm/llvm-project/pull/120314.

**Review by shafik — COMMENTED** [review]
I guess there are no release notes or manual b/c this is not ready to be used by general users on the command line and that will come via another PR?

**Comment by seanm** [comment]
@fhahn can TySan be made to trap when it detects a problem?  I tried `-fsanitize-trap=type` but got:

`clang++: error: unsupported argument 'type' to option '-fsanitize-trap='`

`-fsanitize-trap=all` does not seem to result in TySan trapping, only logging...

**Pull request merged**